File: Labs/Lab 7/Lab7.java
public class lab7 {
    int n = 4;
    int[][] edges = {{0, 1, 3}, {1, 2, 1}, {1, 3, 4}, {2, 3, 1}}; // using test case from lab 7 page
    int distanceThreshold = 4;
    lab7 solution = new lab7();
    int result = solution.findTheCity(n, edges, distanceThreshold);
    System.out.println("Answer: " + result);
}
public int findTheCity(int n, int[][] edges, int distanceThreshold) {
    int[][] dist = new int[n][n];

    for (int i = 0; i < n; i++) {
        Arrays.fill(dist[i], Integer.MAX_VALUE);
        dist[i][i] = 0;  
    }

    for (int[] edge : edges) {
        int from = edge[0];
        int to = edge[1];
        int weight = edge[2];
        dist[from][to] = weight;
        dist[to][from] = weight;  
    }

    for (int k = 0; k < n; k++) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (dist[i][k] != Integer.MAX_VALUE && dist[k][j] != Integer.MAX_VALUE) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
    }

    int minCities = Integer.MAX_VALUE;
    int minCity = -1;
    for (int i = 0; i < n; i++) {
        int reachableCities = 0;
        for (int j = 0; j < n; j++) {
            if (i != j && dist[i][j] <= distanceThreshold) {
                reachableCities++;
            }
        }
        if (reachableCities <= minCities) {
            minCities = reachableCities;
            minCity = i;
        }
    }

    return minCity;

}

File: Labs/Lab 7/Lab7.py
from typing import List

class Lab7:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        dist = [[float('inf')] * n for _ in range(n)]
        for i in range(n):
            dist[i][i] = 0  

        for from_, to, weight in edges:
            dist[from_][to] = weight
            dist[to][from_] = weight  


        for k in range(n):
            for i in range(n):
                for j in range(n):
                    if dist[i][k] != float('inf') and dist[k][j] != float('inf'):
                        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

        min_cities = float('inf')
        min_city = -1

        for i in range(n):
            reachable_cities = 0
            for j in range(n):
                if i != j and dist[i][j] <= distanceThreshold:
                    reachable_cities += 1
            if reachable_cities <= min_cities:
                min_cities = reachable_cities
                min_city = i

        return min_city

n = 4
edges = [[0, 1, 3], [1, 2, 1], [1, 3, 4], [2, 3, 1]]
distanceThreshold = 4
lab7 = Lab7()
result = Lab7.findTheCity(n, edges, distanceThreshold)
print("Answer:", result)
